### ISSUE TITLE
S3 module: Make bucket deletion a more explicit command

### ISSUE THREAD

**Issue by amblina** [ansible/ansible#21796]
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
s3

##### ANSIBLE VERSION
```
ansible 2.2.1.0
  config file = 
  configured module search path = Default w/o overrides
```

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
Local: OSX
Remote: Ubuntu 14

##### SUMMARY
The ``s3`` module parameter ``mode=delete`` for deleting bucket is conceptually similar to ``mode=delobj``.  Prevent users from deleting buckets rather than objects by making ``object`` parameter and ``mode=delobj`` mutually exclusive in task.  Alternatively in the future replace ``mode=delete`` with a more explicit ``mode=delete_bucket``.  This is a particularly important distinction as S3 buckets are, by default, without versioning - the accidental removal of a bucket is something to be avoided.

##### STEPS TO REPRODUCE / CHANGES SUGGESTED

1.  Make ``object`` parameter and ``mode=delobj`` mutually exclusive parameters
2.  Alias then replace ``mode=delete`` with ``mode=delete_bucket``  

```yaml
- name: Delete s3 bucket rather than object
  s3:
    bucket="{{ s3_bucket }}"
    object="{{ s3_path }}/file.tar.gz"
    mode=delete
    region="{{ region }}"

- name: Delete s3 object rather than bucket
  s3:
    bucket="{{ s3_bucket }}"
    object="{{ s3_path }}/file.tar.gz"
    mode=delobj
    region="{{ region }}"

# Proposed change
- name: Delete s3 bucket rather than object
  s3:
    bucket="{{ s3_bucket }}"
    mode=delete   # or alternatively "delete_bucket"
    region="{{ region }}"

- name: Delete s3 object rather than bucket
  s3:
    bucket="{{ s3_bucket }}"
    object="{{ s3_path }}/file.tar.gz"
    mode=delobj
    region="{{ region }}"
```

I would also suggest changing ``mode=delete`` to something more explicit like ``mode=delete_bucket`` in future, but at least preventing the object parameter might make things more explicit than it is now.

**Comment by graingert** [comment]
```yaml
- s3_bucket:
    name: mys3bucket
    state: absent
    force: yes
```

should be the only way to delete a bucket in Ansible.

**Comment by amblina** [comment]
@graingert I completely agree, it would make more sense for creation/deletion of a bucket be kept within s3_bucket rather than s3.  The current setup is ambiguous.

**Comment by graingert** [comment]
The s3 module has an interface totally unlike any other in Ansible, and should be made to be more like the [file](https://docs.ansible.com/ansible/file_module.html) module:
```yaml
- s3:
  bucket: mys3bucket
  path: file.tgz  # or name, or dest
  state: file # or absent etc.
  src: ...
```

however it looks like the s3 module is stable:

> This module is flagged as stableinterface which means that the maintainers for this module guarantee that no backward incompatible interface changes will be made.

so I think there should be a new module, modelled after [file](https://docs.ansible.com/ansible/file_module.html) module:

```yaml
- s3_obj:
  bucket: mys3bucket
  path: file.tgz  # or name, or dest
  state: file # or absent etc.
  src: ...
```

and the s3 module should be deprecated.

This doesn't preclude  #21830  from landing, however, as it's only adding an error case for a redundant option.

**Comment by s-hertel** [comment]
Patches are always welcome! Unfortunately, since it is stable we can't change the interface because that will break playbooks and make people sad. You may want to check out the [s3_sync module](https://docs.ansible.com/ansible/s3_sync_module.html) since it only operates on objects and can't delete buckets.

**Comment by jonatanblue** [comment]
@s-hertel The current version also lets them accidentally delete S3 buckets, which arguably would make them more sad.

**Comment by ryansb** [comment]
The current behavior is documented, and changing the API would break existing playbooks which is against the `stableinterface` definition. The `s3_sync` module is also available, and performs many of the same operations on objects with an eye towards batch operations on many files/objects. That module operates only on objects, so would avoid any chance of accidental bucket operations. 

The current interface is less-than-perfect, but it's also been in since 1.x, and can't just be changed.

**Comment by graingert** [comment]
Is it against policy to raise an error if an object key is included in the config? That's almost certainly an error

**Comment by amblina** [comment]
In which case, perhaps a warning should be added to the documentation at the very least, perhaps directing users to use the sync module and warning that the delete mode deletes buckets containing objects?  I suggest this because AWS CLI doesn't allow for the deletion of buckets which contain objects whereas this module does and versioning your bucket can't save you, it's a pretty dangerous action.

**Comment by s-hertel** [comment]
amblina: You're right, documentation for the distinction between the AWS CLI delete bucket and the s3 delete bucket isn't good, and that would be an easy trap to fall into. I'll make a PR to include a warning in the documentation. Want to check it out and see if it looks okay?

**Comment by amblina** [comment]
@s-hertel One thing to bare in mind with recommendation to use s3_sync module is that according to docs it's available in 2.3 and the current stable release is 2.2.

**Comment by s-hertel** [comment]
@amblina That is true. Bearing that in mind, I guess a warning will be best to implement. Do you have any suggestions for my PR (https://github.com/ansible/ansible/pull/21907/files#diff-68ba194b6312d1232756a1ecb586474bR95) or does it look sufficient?
